source/common/tls/context_impl.cc

```diff
@@ -1344,6 +1394,13 @@ ServerContextImpl::findTlsContext(absl::string_view sni, bool client_ecdsa_capab
 
 enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
+  if (tls_context_provider_factory_cb_ != nullptr) {
+    if (tls_context_provider_ == nullptr) {
+      tls_context_provider_ = tls_context_provider_factory_cb_(shared_from_this());
```

Comment:
I find it confusing that sometimes this points back to itself, and sometimes points to another object. Would it work to have the "normal" code be in a separate object from the ServerContextImpl, and that new object type gets created and set here when there's no config set for this new feature?

Reply:
okay, I'll have a try.